src/vidpf.rs

```diff
@@ -0,0 +1,743 @@
+    }
+
+    /// Panics if weight lengths are different.
+    fn check_eq_length(a: &Self, b: &Self) {
+        if a.0.len() != b.0.len() {
+            panic!("vidpf: weights with different number of field elements")
+        }
+    }
+}
+
+/// [`WeightParameter`] contains the run-time parameters to create an array of field elements.
+#[derive(Debug, Clone, Copy)]
+pub struct WeightParameter<F: FieldElement> {
+    /// The number of field elements.
+    pub length: usize,
+    /// Parameters used to create field elements.
+    pub field_parameter: <F as IdpfValue>::ValueParameter,
+}
+
+impl<F: FieldElement> VidpfWeight for Weight<F> {}
+
+impl<F: FieldElement> IdpfValue for Weight<F> {
+    type ValueParameter = WeightParameter<F>;
+
+    fn generate<S: RngCore>(seed_stream: &mut S, parameter: &Self::ValueParameter) -> Self {
+        Self(
+            (0..parameter.length)
+                .map(|_| <F as IdpfValue>::generate(seed_stream, &parameter.field_parameter))
+                .collect(),
+        )
+    }
+
+    fn zero(parameter: &Self::ValueParameter) -> Self {
+        Self(
+            (0..parameter.length)
+                .map(|_| <F as IdpfValue>::zero(&parameter.field_parameter))
+                .collect(),
+        )
+    }
+
+    fn conditional_select(lhs: &Self, rhs: &Self, choice: Choice) -> Self {
+        Self::check_eq_length(lhs, rhs);
+        Self(
+            zip(&lhs.0, &rhs.0)
+                .map(|(a, b)| <F as IdpfValue>::conditional_select(a, b, choice))
+                .collect(),
+        )
+    }
+}
+
+impl<F: FieldElement> ConditionallyNegatable for Weight<F> {
+    fn conditional_negate(&mut self, choice: Choice) {
+        self.0.iter_mut().for_each(|a| a.conditional_negate(choice));
+    }
+}
+
+impl<F: FieldElement> Add for Weight<F> {
+    type Output = Self;
+
+    /// Panics if weight lengths are different.
+    fn add(self, rhs: Self) -> Self::Output {
+        self.add(&rhs)
+    }
+}
+
+impl<'a, F: FieldElement> Add<&'a Self> for Weight<F> {
+    type Output = Self;
+
+    /// Panics if weight lengths are different.
+    fn add(self, rhs: &Self) -> Self::Output {
+        Self::check_eq_length(&self, rhs);
+        Self(zip(self.0, &rhs.0).map(|(a, b)| a.add(*b)).collect())
+    }
+}
+
+impl<F: FieldElement> AddAssign for Weight<F> {
+    /// Panics if weight lengths are different.
+    fn add_assign(&mut self, rhs: Self) {
+        Self::check_eq_length(self, &rhs);
+        zip(&mut self.0, rhs.0).for_each(|(a, b)| a.add_assign(b));
+    }
+}
+
+impl<F: FieldElement> Sub for Weight<F> {
+    type Output = Self;
+
+    /// Panics if weight lengths are different.
+    fn sub(self, rhs: Self) -> Self::Output {
+        Self::check_eq_length(&self, &rhs);
+        Self(zip(self.0, rhs.0).map(|(a, b)| a.sub(b)).collect())
+    }
+}
+
+impl<F: FieldElement> Encode for Weight<F> {
+    fn encode(&self, bytes: &mut Vec<u8>) -> Result<(), CodecError> {
+        encode_u32_items(bytes, &(), &self.0)
```

Comment:
Remove the length prefix.

For decoding we can provide the expected length as a decoding parameter.

```rust
impl<F: FieldElement> ParameterizedDecode<usize> for Weight<F> {
    fn decode_with_param(length: usize, bytes: &mut Cursor<&[u8]>) -> Result<Self, CodecError> {
        todo!()
    }
}
```

Reply:
good idea, however, removing the prefix makes impossible to implement `Decode` which is required by `IdpfValue` trait. -- unless I'm missing something.

Also the blanket implementation for `ParameterizedDecode for T where T:Decode`  avoids to define another implementation.

* Tests are failing for codec:Encode/Decode.